README.md
# sbcms

NEWS
smbcms v0.99.0 (Release date: 2019-09-06)
=========================
  Initial release!

tests/testthat.R
library(testthat)
library(sbcms)

test_check("sbcms")


tests/testthat/test-QCRSC.R
context ("Test QCRSC function")

test_that ("QC-RSC returns expected output", {

  classes <- sbcdata$class
  batch <- sbcdata$batch
  order <- c(1:nrow(sbcdata$data))
  data <- t(sbcdata$data[, 1:20])

  out <- QCRSC(df = data, order = order, batch = batch, classes = classes,
                  spar = 0, minQC = 4)

  expect_equal (out, testData$QCRSC)

})

test_that ("QC-RSC returns error if QC sample label is wrong", {
  
  classes <- sbcdata$class
  classes[classes == "QC"] <- "Quality"
  batch <- sbcdata$batch
  order <- c(1:nrow(sbcdata$data))
  data <- t(sbcdata$data[, 1:20])
  
  expect_error(QCRSC(df = data, order = order, batch = batch, classes = classes,
               spar = 0, minQC = 4))
})


tests/testthat/test-sbcmsPlot.R
context ("Test sbcmsPlot function")

test_that ("sbcmsPlot returns ggplot output", {

  classes <- sbcdata$class
  batch <- sbcdata$batch
  order <- c(1:nrow(sbcdata$data))
  data <- t(sbcdata$data[, 1:10])

  out <- QCRSC(df = data, order = order, batch = batch, classes = classes,
    spar = 0, minQC = 4)

  plots <- sbcmsPlot(df = data, corrected_df = out, 
    classes, batch, output=NULL, indexes=c(1,5,7))
  
  testthat::expect_s3_class(plots[[1]], "ggplot")
  # Empty plots are getting removed from the list object
  testthat::expect_true(length(plots) == 3)

})

test_that ("sbcmsPlot returns pdf output", {
  
  classes <- sbcdata$class
  batch <- sbcdata$batch
  order <- c(1:nrow(sbcdata$data))
  data <- t(sbcdata$data[, 1:2])
  
  out <- QCRSC(df = data, order = order, batch = batch, classes = classes,
    spar = 0, minQC = 4)
  pdf_output <- tempfile()
  testthat::expect_warning(sbcmsPlot (df = data, corrected_df = out, classes, 
    batch, output=pdf_output))
  
  testthat::expect_true(file.exists(pdf_output))
  testthat::expect_true(file.remove(pdf_output))
})

test_that ("sbcmsPlot returns output for the first 100 features if
  indexes=NULL", {
  
  classes <- sbcdata$class
  batch <- sbcdata$batch
  order <- c(1:nrow(sbcdata$data))
  data <- t(sbcdata$data[, 1:101])
  
  out <- QCRSC(df = data, order = order, batch = batch, classes = classes,
               spar = 0, minQC = 4)
  
  plots <- sbcmsPlot (df = data, corrected_df = out, classes, batch, output=NULL)
  
  testthat::expect_true(length(plots) == 100)
})

test_that ("sbcmsPlot returns ggplot output, with different QC sample label", {
  
  classes <- sbcdata$class
  classes[classes == "QC"] <- "Quality"
  batch <- sbcdata$batch
  order <- c(1:nrow(sbcdata$data))
  data <- t(sbcdata$data[, 1:10])
  
  out <- QCRSC(df = data, order = order, batch = batch, classes = classes,
    spar = 0, minQC = 4, qc_label="Quality")
  
  plots <- sbcmsPlot(df = data, corrected_df = out, 
    classes, batch, output=NULL, indexes=c(1,5,7), qc_label="Quality")
  
  testthat::expect_s3_class(plots[[1]], "ggplot")
  # Empty plots are getting removed from the list object
  testthat::expect_true(length(plots) == 3)
  
})


R/sbc_main.R
#' @importFrom stats median
NULL

##' @title Quality Control-Robust Spline Correction (QC-RSC)
##' @description Implementation of signal correction algorithm as described
##'in Kirwan et al, Anal. Bioanal. Chem., 405 (15), 2013
##' @author Andris Jankevics \email{a.jankevics@bham.ac.uk}
##' @references Kirwan et al, Anal. Bioanal. Chem., 405 (15), 2013
##'\url{https://dx.doi.org/10.1007/s00216-013-6856-7}
##' @details
##' The smoothing parameter (spar) can be optimised using leave-one-out cross
##'validation to avoid overfitting.
##' 
##' @param df A data frame of values to be corrected (samples in columns and
##'features in rows).
##' @param order A numeric vector indicating the order in which samples
##'were measured.
##' @param batch A vector indicating the batch each sample was measured in.
##'If only one batch then all values should be set to 1.
##' @param classes A factor or character vector of sample classes. All QC
##'samples should be labelled  'QC'.
##' @param spar Spline smoothing parameter. Should be in the range 0 to 1.
##'If set to 0 it will be estimated using leave-one-out cross-validation.
##' @param log TRUE or FALSE to perform the signal correction fit on the log
##'scaled data. Default is TRUE.
##' @param minQC Minimum number of QC samples required for signal correction.
##' @param qc_label Class label for QC sample.
##'
##' @return A data frame of corrected values.
##'
##' @examples 
##' classes <- sbcdata$class
##' batch <- sbcdata$batch
##' order <- c(1:nrow(sbcdata$data))
##' data <- t(sbcdata$data[, 1:20])
##' out <- QCRSC(df = data, order = order, batch = batch, classes = classes,
##'spar = 0, minQC = 4)
##' 
##' @export

QCRSC <- function(df, order, batch, classes, spar = 0, log = TRUE,
    minQC = 5, qc_label="QC") {
    
    if (length(which(classes == qc_label)) <= 0) {
        message("QC samples are not defined! Please see help page for
            parameter qc_label.")
        stop()
    }
    
    message("The number of NA and <= 0 values in peaksData before QC-RSC: ",
        sum(is.na(df) | df <= 0))
    
    qcData <- df[, classes == qc_label]
    qc_batch <- batch[classes == qc_label]
    qc_order <- order[classes == qc_label]
    QC_fit <- lapply(seq_len(nrow(df)), sbcWrapper, qcData = qcData, 
        order = order, qcBatch = qc_batch, qcOrder = qc_order, 
        log = log, spar = spar, batch = batch, minQC = minQC)
    QC_fit <- do.call(rbind, QC_fit)
    
    ## Median value for each fature, and divide it by predicted value
    mpa <- apply(df, 1, median, na.rm = TRUE)
    QC_fit <- QC_fit/mpa
    
    ## Divide measured value by correction factor
    res <- df/QC_fit
    res[res <= 0] <- NA
    res
}


R/sbc_methods.R
#' @importFrom stats smooth.spline predict
#' @importFrom utils capture.output
NULL

##' Transform data using natural logarithm
##'
##' @param x Values to convert.
##' @param a log offset.
##' @param inverse To apply log or reverse back to original values.
##' 
##' @return vector of log transformed values

logFit <- function (x, a=1, inverse=FALSE) {
    if (inverse) {
        out <- 0.25 * exp(-x) * (4 * exp(2 * x) - (a * a))
    }else{
        out <- log((x + sqrt(x^2 + a^2))/2)
    }
    return(out)
}

##' Cubic smoothing spline fitting function for batch effect correction
##'
##' @param x QC sample measurement order indices.
##' @param y Signal intensities.
##' @param newX Range of sample indices to predict.
##' @param log TRUE or FALSE to perform the signal correction fit on the log
##'scaled data. Default is TRUE.
##' @param a log scaling offset.
##' @param spar Smoothing parameter of the fitted curve. Should be in the range
##'0 to 1. If set to 0 it will be estimated using leave-one-out
##'cross-validation.
##' @param verbose If set to TRUE will return messages from smooth.spline 
##'fit funtion
##' @return vector of cubic smoothing spline fitted values

splineSmoother <- function(x, y, newX, log=log, a=1, spar, verbose=FALSE) {
    if(log == TRUE) {
        y <- logFit(y, a=a)
    }
    if (spar == 0) {
        # Supress spline fitting CV messages cluttering terminal output
        smoothSplineMessages <- capture.output(sp.obj 
            <- smooth.spline(x, y, cv=TRUE), file=NULL, type="message")
    } else {
        smoothSplineMessages <- character()
        sp.obj <- smooth.spline(x, y, spar=spar)
    }

    valuePredict <- predict(sp.obj, newX)

    if(log==TRUE) {
        valuePredict$y <- logFit(valuePredict$y, a=a, inverse=TRUE)
    }
    
    if (verbose){
        message(paste(smoothSplineMessages, collapse= " || "))
    }
    
    return(valuePredict$y)
}


##' Signal batch correction wrapper function
##'
##' @param id Feature id.
##' @param qcData Data frame with QC samples.
##' @param qcBatch QC batch numbers.
##' @param qcOrder QC measurement order.
##' @param spar Spline smoothing parameter. Should be in the range 0 to 1.
##'If set to 0 it will be estimated using leave-one-out cross-validation.
##' @param log TRUE or FALSE to perform the signal correction fit on the
##'log scaled data. Default is TRUE.
##' @param batch A vector indicating the batch each sample was measured in.
##'If only one batch was measured then all values should be set to 1.
##' @param minQC Minimum number of QC samples required for signal correction.
##' @param order A numeric vector indicating the order in which samples
##'were measured.
##' @return vector of corrected values of selected feature for QC data

sbcWrapper <- function(id, qcData, order, qcBatch, qcOrder, log=log, spar=spar,
    batch=batch, minQC) {
    
    out <- tryCatch ({
    # Measurment order can be non consecutive numbers as well
    maxOrder <- length(order)
    newOrder <- seq_len(maxOrder)
    subData <- qcData[id, ]
    nbatch <- unique(qcBatch)

    outl <- matrix(nrow=maxOrder, ncol=length(nbatch))

    for (nb in seq_len(length(nbatch))) {
        x <- qcOrder[qcBatch == nbatch[nb]]
        y <- subData[qcBatch == nbatch[nb]]
        NAhits <- which(is.na(y))

        if (length(NAhits) > 0) {
            x <- x[-c(NAhits)]
            y <- y[-c(NAhits)]
        }

        if (length(y) >= minQC) {
            outl[,nb] <- splineSmoother(x=x, y=y, newX=newOrder, log=log,
            a=1, spar=spar)
        } else {
            outl[,nb] <- rep(NA, maxOrder)
        }
    }

    outp <- rep(NA, nrow(outl))
    for (nb in seq_len(length(nbatch))) {
        range <- c(batch == nbatch[nb])
        outp[range] <- outl[range, nb]
    }

    outp
    },

    error = function(e){
        print (e, "\n")
        return(NULL)
    })
    return(outp)
}


R/plot_corrected_data.R
#' @import ggplot2
#' @import reshape2
#' @importFrom grDevices pdf dev.off
NULL

##' Plot the output from signal batch correction for the first 100 features
##'
##' @param df A data frame containing the original data before correction
##'  (samples in columns, features in rows).
##' @param corrected_df A data frame containing the corrected data,
##'  as output by QCRSC
##' @param batch A vector indicating the batch each sample was measured in.
##'  If only one batch was measured then all values should be set to 1
##' @param classes A factor or character vector of sample classes.
##'  All QC samples should be labelled "QC".
##' @param order A numeric vector  indicating the order in which samples
##'  were measured.
##' @param output Filename of the output pdf file. Can include the path.
##' 
##' @return Pdf file showing data before and after signal correction
##' 
##' @examples 
##' classes <- sbcdata$class
##' batch <- sbcdata$batch
##' order <- c(1:nrow(sbcdata$data))
##' data <- t(sbcdata$data[, 1:20])
##' out <- QCRSC(df = data, order = order, batch = batch, classes = classes,
##'  spar = 0, minQC = 4)
##' plot_sbc (df = data, corrected_df = out, classes, batch, order,
##'  output="sbcms_plots.pdf") 
##'  
##' @export

plot_sbc <- function (df, corrected_df, classes, batch, order,
  output="sbcms_plots.pdf")
{
  shapes <- rep(1,length(classes))
  shapes[classes == "QC"] <- 3

  manual_color <- c("#386cb0", "#ef3b2c", "#7fc97f", "#fdb462", "#984ea3",
    "#a6cee3", "#778899", "#fb9a99", "#ffff33")

  gg_THEME <- theme(
    panel.background=element_blank(),
    panel.grid.major=element_line(color="gray80", size=0.3),
    axis.line=element_line(color="black"),
    axis.text=element_text(color="black"),
    axis.title=element_text(color="black"),
    panel.grid.minor.x=element_line(color="gray80", size=0.3,
      linetype="dashed"),
    panel.grid.minor.y=element_line(color="gray80", size=0.3)

  )

  plots <- list()

  plotLim <- 100
  if (nrow(df) < 100) plotLim <- nrow(df)

  for (peakn in seq_len(plotLim)){
    cat (peakn, "\n")
    A <- data.frame (x=c(seq_len(ncol(df))), original=log(df[peakn, ],10),
                     fitted=log(corrected_df[peakn, ],10),
                     batch=as.factor(batch), shapes=shapes)
    A <- melt(A, id.vars=c("x","batch","shapes"))

    plots[[peakn]] <- ggplot (A, aes_(~x, ~value, col=~batch, shape=~shapes))+
      facet_grid(variable ~ .)+
      geom_point() +scale_shape_identity()+
      scale_colour_manual(values=manual_color)+
      ggtitle(row.names(df)[peakn])+
      ylab("log10(intensity)")+
      xlab("injection order")+
      gg_THEME
  }

  pdf (output)
  invisible(lapply(plots, print))
  dev.off()

}


R/sbcmsPlot.R
#' @import ggplot2
#' @importFrom grDevices pdf dev.off
#' @importFrom reshape2 melt
NULL

##' Plot the output from signal batch correction for the selected or
##'the first 100 features
##'
##' @param df A data frame containing the original data before correction
##'(samples in columns, features in rows).
##' @param corrected_df A data frame containing the corrected data,
##'as output by QCRSC
##' @param batch A vector indicating the batch each sample was measured in.
##'If only one batch was measured then all values should be set to 1
##' @param classes A factor or character vector of sample classes.
##'All QC samples should be labelled 'QC'.
##' @param indexes Numeric vector defining whihc features from data frame to 
##'plot. If set to NULL will plot the first 100.
##' @param output Filename of the output pdf file. Can include the path.
##'If set to NULL output will be list object containing ggplot2 plots.
##' @param qc_label Class label for QC sample.
##' 
##' @return Pdf file or list object showing data before and after 
##'signal correction
##' 
##' @examples 
##' classes <- sbcdata$class
##' batch <- sbcdata$batch
##' data <- t(sbcdata$data[, 1:20])
##' order <- c(1:ncol(data))
##' out <- QCRSC(d =data, order=order, batch=batch, classes=classes,
##'     spar=0, minQC=4)
##' plots <- sbcmsPlot (df=data, corrected_df=out, classes, batch,
##'     output=NULL) 
##'
##' @export

sbcmsPlot <- function(df, corrected_df, classes, batch, indexes = NULL,
    qc_label="QC", output = "sbcms_plots.pdf") {
    
    shapes <- rep(19, length(classes))
    shapes[classes == qc_label] <- 3
    manual_color <- c("#386cb0", "#ef3b2c", "#7fc97f", "#fdb462", "#984ea3", 
        "#a6cee3", "#778899", "#fb9a99", "#ffff33")
    gg_THEME <- theme(panel.background = element_blank(), 
        panel.grid.major = element_line(color = "gray80", size = 0.3), 
        axis.line = element_line(color = "black"), 
        axis.text = element_text(color = "black"), 
        axis.title = element_text(color = "black"), 
        panel.grid.minor.x = element_line(color = "gray80", 
        size = 0.3, linetype = "dashed"), 
        panel.grid.minor.y = element_line(color = "gray80", size = 0.3))
    plots <- list()
    
    if (is.null(indexes) & nrow(df) >= 100) {
        indexes <- seq_len(100)
    } else if (nrow(df) < 100 & is.null(indexes)) {
        indexes <- seq_len(nrow(df))
    }
    for (peakn in indexes) {
        A <- data.frame(x = c(seq_len(ncol(df))), 
            original = log(df[peakn, ], 10),
            corrected = log(corrected_df[peakn, ], 10), 
            batch = as.factor(batch), shapes = shapes)
        A <- melt(A, id.vars = c("x", "batch", "shapes"))
        
        plots[[peakn]] <- ggplot(A, 
            aes_(~x, ~value, col = ~batch, shape = ~shapes)) + 
            facet_grid(variable ~ .) + geom_point() + 
            scale_shape_identity() + geom_point(size=2) +
            scale_colour_manual(values = manual_color) + 
            ggtitle(row.names(df)[peakn]) + ylab("log10(intensity)") + 
            xlab("injection order") + gg_THEME
    }
    # remove lists with NULL values
    plots <- plots[lengths(plots) != 0]
    
    if (!is.null(output)) {
        pdf(output)
        invisible(lapply(plots, print))
        dev.off()
    } else {
        return(plots)
    }
}
